Sources/AWSLambdaRuntimeCore/HTTPClient.swift

```diff
@@ -285,14 +236,19 @@ private final class UnaryHandler: ChannelDuplexHandler {
         //
         // That's why we only assume the connection shall be closed if we receive
         // a "connection = close" header.
-        let serverCloseConnection = response.headers.first(name: "connection")?.lowercased() == "close"
+        let serverCloseConnection =
+            response.head.headers["connection"].contains(where: { $0.lowercased() == "close" })
 
-        if !self.keepAlive || serverCloseConnection || response.version != .init(major: 1, minor: 1) {
-            pending.promise.futureResult.whenComplete { _ in
-                _ = context.channel.close()
-            }
+        let closeConnection = serverCloseConnection || response.head.version != .http1_1
+
+        if closeConnection {
+            self.state = .waitForConnectionClose(httpResponse, promise)
+            _ = context.channel.close()
+            return
```

Comment:
This is a rather interesting workaround:

If we succeed the promise here and close the connection after succeeding the promise we may run into a race condition:

The lambda runtime will ask for the next work item directly after a succeeded post response request. The desire for the next work item might be faster than the attempt to close the connection. This will lead to a situation where we close the connection but the next request has already been scheduled on the connection that we want to close. For this reason we postpone succeeding the promise until the connection has been closed. This codepath will only be hit in the very, very unlikely event of the server demanding to close connection. (It's more or less only implemented to support http1.1 correctly.) Correct behavior is ensured with the test `testNoKeepAliveServer`.